### PR TITLE
Refactor and improve `NamedTuple` deserialization from JSON and YAML

### PR DESCRIPTION
**Pull request by HertzDevil**
* Makes the temporary value variables non-nilable unless their corresponding types are nilable. This saves the downcasts at the end of deserialization.
* Fixes compilation error for deserialization of `NamedTuple`s that use file-private value types.
* ~~Fixes compilation error for deserialization of empty `NamedTuple`s.~~

### CODE REVIEW

File: src/json/from_json.cr [hunk 1]
```diff
@@ -266,9 +266,16 @@ end
...
-      %var{key.id} = nil
-      %found{key.id} = false
+    {% for key, type in T %}
+      {% if type.nilable? %}
+        %var{key.id} = nil
+      {% else %}
+        %var{key.id} = uninitialized typeof(begin
+          value = uninitialized self
+          value[{{ key.symbolize }}]
+        end)
```
Comment (straight-shoota) [inline src/json/from_json.cr:276]: Wouldn't this work?
```suggestion
        %var{key.id} = uninitialized {{ type }}
```

**Review by straight-shoota — COMMENTED** [review]
I'm wondering if it wouldn't be easier to always use an uninitailized value and keep track if it was found. And only apply nil handling for nilable properties when constructing the final tuple.

File: src/json/from_json.cr (hunk 1 above)
Comment (asterite) [reply src/json/from_json.cr:276]: I think we should put a comment around why we do it this way. I think I made the exact same comment in a previous PR but now I can't remember why doing it the way @straight-shoota (and previously me) suggest is not a good idea.

Comment (HertzDevil) [reply src/json/from_json.cr:276]: It's #11834. But there is an even more idiomatic way of expressing this: #12011

**Review by sdogruyol — APPROVED** [review]
Thank you @HertzDevil 🙏

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**